deep_qa/training/models.py

```diff
@@ -146,6 +151,183 @@ def _make_predict_function(self):
+            elif shuffle:
+                numpy.random.shuffle(index_array)
+
+            batches = _make_batches(num_train_samples, batch_size)
+            epoch_logs = {}
+            for batch_index, (batch_start, batch_end) in enumerate(batches):
+                batch_ids = index_array[batch_start:batch_end]
+                try:
+                    if isinstance(ins[-1], float):
+                        # Do not slice the training phase flag.
+                        ins_batch = _slice_arrays(ins[:-1], batch_ids) + [ins[-1]]
+                    else:
+                        ins_batch = _slice_arrays(ins, batch_ids)
+                except TypeError:
+                    raise TypeError('TypeError while preparing batch. '
+                                    'If using HDF5 input data, '
+                                    'pass shuffle="batch".')
+
+                # Here is the main difference between a single gpu model and one split
+                # across multiple gpus. In our multiple gpu model, all of the inputs
+                # are replicated num_gpus times, so we need to split our large batch
+                # into the corresponding sets of smaller batches for each model.
+                if hasattr(self, "num_gpus"):
+
+                    def multi_gpu_batch(variable_list):
+                        # Splits up and orders a list of inputs for a single
+                        # model into
+                        split_batch = slice_batch(variable_list, self.num_gpus)  # pylint: disable=no-member
+                        ordered_var_list = []
+                        for single_model_variables in zip(*split_batch):
+                            ordered_var_list.extend(single_model_variables)
+                        return ordered_var_list
+
+                    if isinstance(ins_batch[-1], float):
+                        model_inputs = multi_gpu_batch(ins_batch[:-1])
+                        model_inputs.append(ins_batch[-1])
+                    else:
+                        model_inputs = multi_gpu_batch(ins_batch)
+                    ins_batch = model_inputs
+
+                batch_logs = {}
+                batch_logs['batch'] = batch_index
+                batch_logs['size'] = len(batch_ids)
+                callbacks.on_batch_begin(batch_index, batch_logs)
+                outs = f(ins_batch)
+                if not isinstance(outs, list):
+                    outs = [outs]
+                for label, output in zip(out_labels, outs):
+                    batch_logs[label] = output
+
+                callbacks.on_batch_end(batch_index, batch_logs)
+
+                if batch_index == len(batches) - 1:  # Last batch.
+                    if do_validation:
+                        # If we are using multiple gpus, our batch size will be
+                        # scaled up accordingly. However, validation will run
+                        # on a single gpu, so we divide by the number of gpus
```

Comment:
This requires that we also override the `make_test_function` to provide multi-gpu functionality. I think it would be better to keep multi-gpu stuff as training only for the time being, as this is the time intensive part - we can add it later if required.